pkg/beholder/client_test.go

```diff
@@ -0,0 +1,345 @@
+package beholder
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploggrpc"
+	otellog "go.opentelemetry.io/otel/log"
+	otelglobal "go.opentelemetry.io/otel/log/global"
+	otelmetric "go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/propagation"
+	sdklog "go.opentelemetry.io/otel/sdk/log"
+	oteltrace "go.opentelemetry.io/otel/trace"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/beholder/internal/mocks"
+)
+
+type MockExporter struct {
+	mock.Mock
+	sdklog.Exporter
+}
+
+func (m *MockExporter) Export(ctx context.Context, records []sdklog.Record) error {
+	args := m.Called(ctx, records)
+	return args.Error(0)
+}
+
+func (m *MockExporter) Shutdown(ctx context.Context) error {
+	args := m.Called(ctx)
+	return args.Error(0)
+}
+
+func (m *MockExporter) ForceFlush(ctx context.Context) error {
+	args := m.Called(ctx)
+	return args.Error(0)
+}
+
+func TestClient(t *testing.T) {
+	defaultCustomAttributes := func() map[string]any {
+		return map[string]any{
+			"int_key_1":            123,
+			"int64_key_1":          int64(123),
+			"int32_key_1":          int32(123),
+			"str_key_1":            "str_val_1",
+			"bool_key_1":           true,
+			"float_key_1":          123.456,
+			"byte_key_1":           []byte("byte_val_1"),
+			"str_slice_key_1":      []string{"str_val_1", "str_val_2"},
+			"nil_key_1":            nil,
+			"beholder_data_schema": "/schemas/ids/1001", // Required field, URI
+		}
+	}
+	defaultMessageBody := []byte("body bytes")
+
+	tests := []struct {
+		name                   string
+		makeCustomAttributes   func() map[string]any
+		messageBody            []byte
+		messageCount           int
+		exporterMockErrorCount int
+		exporterOutputExpected bool
+		messageGenerator       func(client OtelClient, messageBody []byte, customAttributes map[string]any)
+	}{
+		{
+			name:                   "Test Emit",
+			makeCustomAttributes:   defaultCustomAttributes,
+			messageBody:            defaultMessageBody,
+			messageCount:           10,
+			exporterMockErrorCount: 0,
+			exporterOutputExpected: true,
+			messageGenerator: func(client OtelClient, messageBody []byte, customAttributes map[string]any) {
+				err := client.Emitter.Emit(context.Background(), messageBody, customAttributes)
+				assert.NoError(t, err)
+			},
+		}, {
+			name:                   "Test EmitMessage",
+			makeCustomAttributes:   defaultCustomAttributes,
+			messageBody:            defaultMessageBody,
+			messageCount:           10,
+			exporterMockErrorCount: 0,
+			exporterOutputExpected: true,
+			messageGenerator: func(client OtelClient, messageBody []byte, customAttributes map[string]any) {
+				message := NewMessage(messageBody, customAttributes)
+				err := client.Emitter.EmitMessage(context.Background(), message)
```

Comment:
From these tests its hard for me to see the value of both an `Emit` and `EmitMessage` API - would you be able to provide some details on that or perhaps more illustrative testing?

Reply:
@patrickhuie19, 

Please see the related conversion https://github.com/smartcontractkit/chainlink-common/pull/694#discussion_r1714219755

I'm ok with reducing Message Emitter API to a single method `EmitMessage(context.Context, Message) error`


See usage examples:

https://github.com/smartcontractkit/chainlink-common/pull/694/files#diff-d5bbc8388dc6b5c9b3b4011eb5cd2448839eb67b352eae919a8a68fb3c3fdb12R117-R120

https://github.com/smartcontractkit/chainlink-common/pull/694/files#diff-d5bbc8388dc6b5c9b3b4011eb5cd2448839eb67b352eae919a8a68fb3c3fdb12R144-R147

https://github.com/smartcontractkit/atlas/blob/master/beholder/cmd/demo/main.go#L31-L46

Reply:
I'm in favor of starting with the lowest surface area API. If we need to add more later on, we can, but I'd like to err on the side of supporting as few exported methods as possible for backwards compatibility. 

I believe an `Emit` API that takes in key values and will construct the message for you is the cleanest developer experience

Reply:
Makes sense